### ISSUE TITLE
Automate running label-approved for nixpkgs PRs

### ISSUE THREAD

**Issue by delroth** [NixOS/nixos-org-configurations#267]
**Is your feature request related to a problem? Please describe.**

https://github.com/nix-community/label-approved is a useful script which labels PRs based on the number of approvals they have received as well as whether maintainers of impacted derivations have approved the PR. Those labels can be extremely useful for mergers like me to find PRs that have been reviewed and approved by maintainers, but haven't been merged for some reason.

Right now the labels are applied manually or when a nixpkgs committer runs the automation script themselves (not sure whether that actually happens?). I suspect that widespread availability of those labels could help reduce PR merge time, and would significantly help maintainers that don't have merge privileges.

cc @artturin, author of the automation in question. Would love your input.
cc @RaitoBezarius @lilyinstarlight who seemed interested when I raised this proposal on Matrix.

**Describe the solution you'd like**

Similar to https://github.com/NixOS/nixos-org-configurations/pull/264 but for https://github.com/nix-community/label-approved :)

**Describe alternatives you've considered**

This might be implementable as a GitHub action, which might lead to fewer issues with GH API quota. However nobody has written an implementation of this yet to prove that it's actually doable.

**Additional context**

Relevant Matrix discussion on #dev:nixos.org, 2023-08-23 10:33 GMT (but this issue captures a lot of what was said in this discussion).

**Comment by RaitoBezarius** [comment]
cc @JulienMalka (krkrkrkr)

**Comment by JulienMalka** [comment]
Hello, this seems to be a good use of the non-critical infrastructure, where we are currently deploying a first-time-contribution tagger. If you feel like doing it, can you open a PR in this repo and the subfolder "non-critical-infra" to deploy this ? Otherwise, I'll take care of it but that may take a bit longer :)

**Comment by delroth** [comment]
I can probably find some time in the next few days to get this done. As @mweinelt mentioned on IRC, this is likely however going to be blocked by secret management (since it requires a GH API key to operate).

**Comment by JulienMalka** [comment]
No problem, I can handle the secret management part :)

**Comment by delroth** [comment]
I've been running this on my own infra right now to ensure it works fine, and bad news: it's really annoying to keep it from hitting into GH rate limits (which is something other people suggested might be an issue -- but I was hoping not :p)

This should probably be either: 1. rearchitectured to run as a GH action; 2. rearchitectured to run as a webhook receiver (does NixOS's infra already have a webhook receiver somewhere? this would be a fairly trivial addition to an existing one).

**Comment by mweinelt** [comment]
I had the impression, that only the initial run was expensive and it would keep state, so it wouldn't have to iterate over everything again.

@Janik-Haag Can you commment on that?

**Comment by delroth** [comment]
You're thinking of the other bot that was recently deployed :)

**Comment by lilyinstarlight** [comment]
Fwiw I think this tool would be better done as a GitHub Action that triggers on `pull_request_review` and `push` (as @delroth seems to be implementing) rather than a period runner on non-critical-infra here anyway

**Comment by Janik-Haag** [comment]
Yeah I think lily is right, since there is no really efficient way you can do it otherwise because you have to periodically check a bunch of stuff otherwise the only way I can think of optimizing some stuff is using githubs graphql api for some querrys since it is has it's own api limit and is way more granular.

And I think you are thinking of the first-time-contribution stuff @mweinelt ? And yes that is only expensive the first time since it has to work through every existing pr but after that it just caches the pr number where it left of and all the authors it has encountered.

**Comment by delroth** [comment]
Superseded by https://github.com/NixOS/nixpkgs/pull/264633
